presto-main/src/main/java/com/facebook/presto/dispatcher/LocalDispatchQuery.java

```diff
@@ -97,6 +104,63 @@ public LocalDispatchQuery(
         });
     }
 
+    @Override
+    public void startWaitingForPrerequisites()
+    {
+        // It's possible that queryExecution fails before we start for prerequisites, in that case, don't even
+        // start waiting for prerequisites
+        if (isDone()) {
+            return;
+        }
+
+        try {
+            Session session = stateMachine.getSession();
+            CompletableFuture<?> prerequisitesFuture = queryPrerequisites.waitForPrerequisites(
+                    stateMachine.getQueryId(),
+                    new QueryPrerequisitesContext(
+                            session.getCatalog(),
+                            session.getSchema(),
+                            stateMachine.getBasicQueryInfo(Optional.empty()).getQuery(),
+                            session.getSystemProperties(),
+                            session.getConnectorProperties()));
+
+            addStateChangeListener(state -> {
+                if (state.isDone()) {
+                    queryPrerequisites.queryFinished(stateMachine.getQueryId());
+                    if (!prerequisitesFuture.isDone()) {
+                        prerequisitesFuture.cancel(true);
+                    }
+                }
+            });
+
+            prerequisitesFuture.whenCompleteAsync((result, throwable) -> {
```

Comment:
I think we should add a timeout here to be defensive against a bad implementation that takes too long to complete. We could cancel the thread from here, rather than relying on implementations to properly enforce their own timeouts and cancel themselves.

Reply:
We already have the session properties `query_max_run_time` which controls how long can a query stay in the system. And if that is reached - the query will fail and that will in turn cancel the future. What extra value do you feel can adding an additional timeout here add ? The implementations can always take care of this themselves by ensuring that they add a timeout on their end if they want a specific timeout